File: src/main/java/nihilite/agent/Worker.java
package nihilite.agent;

import clojure.java.api.Clojure;
import clojure.lang.Compiler;
import clojure.lang.DynamicClassLoader;
import clojure.lang.IFn;
import clojure.lang.Symbol;

import java.util.logging.Level;
import java.util.logging.Logger;

public final class Worker implements Runnable {

    private static final Logger LOG = Logger.getLogger("nihilite.agent.Worker");

    @Override
    public void run() {
        try {
            initClojure();
            bindCompilerLoader();
        } catch (Throwable t) {
            LOG.log(Level.SEVERE, "[Nihilite-agent] worker failed", t);
        } finally {
            nihilite.agent.Agent.signalWorkerReady();
        }
    }

    private static void initClojure() {
        IFn require = Clojure.var("clojure.core", "require");
        require.invoke(Symbol.intern("clojure.core"));
        require.invoke(Symbol.intern("nihilite.transport"));
        require.invoke(Symbol.intern("nihilite.boot"));
        require.invoke(Symbol.intern("nihilite.registry"));

        try {
            IFn installRedisp = Clojure.var(
                    "nihilite.registry", "install-redefine-dispatcher!");
            Object result = installRedisp.invoke();
            LOG.info("[Nihilite-agent] redefine dispatcher installed: " + result);
        } catch (Throwable t) {
            LOG.log(Level.WARNING,
                    "[Nihilite-agent] install-redefine-dispatcher! failed; "
                            + ":redefine specs will throw at first call",
                    t);
        }

        try {
            IFn evalInit = Clojure.var("nihilite.boot", "eval-init!");
            Object initResult = evalInit.invoke();
            LOG.info("[Nihilite-agent] init form evaled: " + initResult);
        } catch (Throwable t) {
            LOG.log(Level.WARNING,
                    "[Nihilite-agent] eval-init! failed; continuing",
                    t);
        }
    }

    private static void bindCompilerLoader() {
        ClassLoader hostCl = resolveHostClassLoader();
        DynamicClassLoader loader = new DynamicClassLoader(hostCl);
        Compiler.LOADER.bindRoot(loader);
        LOG.info(String.format("[Nihilite-agent] Compiler/LOADER bound to %s wrapping %s",
                loader.getClass().getName(), hostCl));
    }

    private static ClassLoader resolveHostClassLoader() {
        String hint = System.getProperty("nihilite.compiler-loader-hint", "");
        if (!hint.isEmpty()) {
            java.lang.instrument.Instrumentation inst =
                    nihilite.agent.Agent.currentInstrumentation();
            if (inst != null) {
                String dotName = hint.replace('/', '.');
                for (Class<?> c : inst.getAllLoadedClasses()) {
                    if (c != null && dotName.equals(c.getName()) && c.getClassLoader() != null) {
                        LOG.info(String.format(
                                "[Nihilite-agent] compiler-loader hint '%s' resolved to %s",
                                hint, c.getClassLoader()));
                        return c.getClassLoader() == null ? ClassLoader.getSystemClassLoader() : c.getClassLoader();
                    }
                }
                LOG.warning(String.format(
                        "[Nihilite-agent] compiler-loader hint '%s' not found among loaded classes; "
                                + "falling back to system classloader", hint));
            } else {
                LOG.warning("[Nihilite-agent] compiler-loader hint set but no Instrumentation "
                        + "(standalone server mode); falling back to system classloader");
            }
        }
        return ClassLoader.getSystemClassLoader();
    }
}

File: src/main/clojure/nihilite/kernel/worker.clj
(ns nihilite.kernel.worker
  "Background worker thread that brings up Clojure-side state after the
   JVM has invoked premain/agentmain.

   Equivalent to the deleted nihilite.agent.Worker Java class. Runs on
   a dedicated thread started by nihilite.kernel.Agent.startWorkerOnce
   so that the premain call can return to the JVM immediately without
   blocking on Clojure runtime init.

   Note: avoids `(:require [clojure.java.api :as api])` because that
   namespace ships only as `Clojure.class` (no .clj source) in clojure
   1.12.x; requiring the namespace directly fails to find the source
   file at runtime when the project is bundled into a fat jar."
  (:require [clojure.tools.logging :as log])
  (:import [clojure.lang DynamicClassLoader Compiler]))

(defn- clojure-var
  "Returns the IFn for a (ns-name, var-name) pair, without requiring
   `clojure.java.api` at compile- or load-time. Uses clojure.lang.RT/var
   so we don't depend on resolving a Namespace Class via Class/forName
   (which fails when the worker thread's context classloader can't find
   clojure.core)."
  [^String ns-name ^String var-name]
  (clojure.lang.RT/var ns-name var-name))

(defn- require-ns [sym]
  (.invoke ^clojure.lang.IFn (clojure-var "clojure.core" "require") sym))

(defn- init-clojure
  "Loads the core Clojure namespaces and installs the redefine dispatcher
   bridge. Errors are caught and logged so the worker can still complete
   the loader binding below."
  []
  (require-ns 'clojure.core)
  (require-ns 'nihilite.transport)
  (require-ns 'nihilite.boot)
  (require-ns 'nihilite.registry)
  (try
    (let [install-redisp (clojure-var "nihilite.registry" "install-redefine-dispatcher!")
          result (.invoke ^clojure.lang.IFn install-redisp)]
      (log/info "[Nihilite Agent] redefine dispatcher installed:" result))
    (catch Throwable t
      (log/warn t "[Nihilite Agent] install-redefine-dispatcher! failed"))))

(defn- find-hinted-class-loader
  "If `nihilite.compiler-loader-hint` is set and Instrumentation is
   available, returns the classloader of the first loaded class whose
   name matches the hint; otherwise returns nil."
  [hint]
  (let [agent-currentInst (resolve 'nihilite.kernel.agent/agent-currentInstrumentation)]
    (when agent-currentInst
      (let [inst (agent-currentInst)]
        (when inst
          (let [dot-name (.replace ^String hint "/" ".")]
            (try
              (some (fn [^Class c]
                      (when (and c (.equals dot-name (.getName c))
                                 (.getClassLoader c))
                        (.getClassLoader c)))
                    (.getAllLoadedClasses ^java.lang.instrument.Instrumentation inst))
              (catch Throwable t
                (log/warn t "[Nihilite Agent] hint lookup failed")
                nil))))))))

(defn resolve-host-class-loader
  "Returns the classloader the Clojure Compiler should use, honoring the
   `nihilite.compiler-loader-hint` system property when present."
  []
  (let [hint (System/getProperty "nihilite.compiler-loader-hint" "")]
    (if-not (seq hint)
      (ClassLoader/getSystemClassLoader)
      (let [match (find-hinted-class-loader hint)]
        (if match
          (do
            (log/info "[Nihilite Agent] compiler-loader hint '" hint
                      "' resolved to" match)
            match)
          (do
            (log/warn "[Nihilite Agent] compiler-loader hint '" hint
                      "' not found among loaded classes; "
                      "falling back to system classloader")
            (ClassLoader/getSystemClassLoader)))))))

(defn- bind-compiler-loader
  "Replaces clojure.lang.Compiler/LOADER with a DynamicClassLoader that
   wraps the resolved host classloader. Lets `require` resolve user
   classes at runtime via ByteBuddy-instrumented classloaders."
  []
  (let [host-cl (resolve-host-class-loader)
        loader (DynamicClassLoader. host-cl)]
    (.bindRoot Compiler/LOADER loader)
    (log/info "[Nihilite Agent] Compiler/LOADER bound:" (.getName (class loader)))))

(defn init-and-bind
  "Brings Clojure-side state online. Called from the agent worker thread.
   Binds *ns* to this namespace so tools.logging resolves the logger
   name to `nihilite.kernel.worker` instead of `clojure.tools.logging$eval...`."
  []
  (binding [*ns* (find-ns 'nihilite.kernel.worker)]
    (try
      (init-clojure)
      (bind-compiler-loader)
      (catch Throwable t
        (log/error t "[Nihilite Agent] worker failed")))))
